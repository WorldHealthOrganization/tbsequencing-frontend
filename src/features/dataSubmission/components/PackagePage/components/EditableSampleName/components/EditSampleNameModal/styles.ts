import { css } from '@emotion/react';

export const buttonsContainer = css({
  marginTop: 32,
  justifyContent: 'flex-end',
  display: 'flex',
});

export const fieldWrapper = css({
  marginTop: 16,
});

export const okayBtn = css({
  marginLeft: 16,
  justifyContent: 'flex-end',
});

export const fieldsContainer = css({
  marginTop: 32,
  marginBottom: 32,
});
