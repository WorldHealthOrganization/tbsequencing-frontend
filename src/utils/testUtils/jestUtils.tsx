import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
