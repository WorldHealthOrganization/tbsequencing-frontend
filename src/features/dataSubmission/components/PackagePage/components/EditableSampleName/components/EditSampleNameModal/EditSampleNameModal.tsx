/** @jsxImportSource @emotion/react */
import React, { useEffect } from 'react';
import { Control, FieldErrors, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup/dist/yup';
import * as yup from 'yup';
import { toast } from 'react-toastify';
import * as styles from './styles';
import { AbstractIField } from '../../../../../../../auth/components/registration/types';
import { validateTextField } from '../../../../../../../../utils/validators';
import BaseModal from '../../../../../../../../components/BaseModal';
import H2 from '../../../../../../../../components/typography/H2';
import FormFields from '../../../../../../../../components/FormFields';
import { AppButton } from '../../../../../../../../components/AppButton/AppButton';
import { useUpdateSampleMutation } from '../../../../../../../../services/submissionApi/submissionApi';

interface Props {
  name: string;
  open: boolean;
  handleCloseModal: () => void;
  sampleId: number;
  packageId: number;
}

interface IInputs {
  name: string;
}

const fields: AbstractIField<IInputs>[] = [{
  name: 'name',
  type: 'text',
  placeholder: 'Type sample name here',
  label: 'Sample name',
  wrapperStyles: styles.fieldWrapper,
}];

const formSchema = yup.object().shape({
  name: validateTextField('Sample name is required'),
});

export const EditSampleNameModal = ({
  name, open, handleCloseModal, sampleId, packageId,
}: Props) => {
  const defaultValues: IInputs = {
    name,
  };

  const {
    control, formState: { errors, isValid }, handleSubmit, reset,
  } = useForm<IInputs>({
    resolver: yupResolver(formSchema),
    defaultValues,
    mode: 'onChange',
  });

  const [updateSample, { error, data }] = useUpdateSampleMutation();
  const onSubmit = (formData: IInputs) => {
    updateSample({
      name: formData.name, packageId, sampleId,
    });
  };

  useEffect(() => {
    if (data) {
      toast.success('Sample name was successfully updated');
      handleCloseModal();
    }

    if (error) {
      toast.error(
        // @ts-ignore
        error.data.errors[0].detail,
      );
    }
  }, [error, data]);

  const onCancel = () => {
    handleCloseModal();
    reset(defaultValues);
  };

  return (
    <BaseModal open={open}>
      <>
        <H2>Edit sample name</H2>
        <div css={styles.fieldWrapper}>
          <FormFields
            <IInputs>
            errors={errors as FieldErrors}
            loading={false}
            fieldArray={fields}
            control={control as unknown as Control}
          />
        </div>

        <div css={styles.buttonsContainer}>
          <AppButton onClick={onCancel} size="large" variant="outlined">Cancel</AppButton>
          <AppButton
            disabled={!isValid}
            style={styles.okayBtn}
            size="large"
            variant="contained"
            onClick={handleSubmit(onSubmit)}
          >
            Save
          </AppButton>
        </div>
      </>
    </BaseModal>
  );
};

export default EditSampleNameModal;
