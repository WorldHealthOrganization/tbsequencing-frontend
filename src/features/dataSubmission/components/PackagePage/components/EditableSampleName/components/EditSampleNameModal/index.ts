import EditSampleNameModal from './EditSampleNameModal';

export default EditSampleNameModal;
