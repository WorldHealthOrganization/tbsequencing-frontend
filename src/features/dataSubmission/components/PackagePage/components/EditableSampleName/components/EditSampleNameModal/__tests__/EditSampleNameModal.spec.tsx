import React from 'react';
import { render, getTreeJSONWithRedux } from '../../../../../../../../../utils/testUtils/jestUtils';
import EditSampleNameModal from '../EditSampleNameModal';

describe('EditSampleNameModal component test', () => {
  test('renders EditSampleNameModal component without error', () => {
    render(
      <EditSampleNameModal />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <EditSampleNameModal />,
    );

    expect(tree).toMatchSnapshot();
  });
});
