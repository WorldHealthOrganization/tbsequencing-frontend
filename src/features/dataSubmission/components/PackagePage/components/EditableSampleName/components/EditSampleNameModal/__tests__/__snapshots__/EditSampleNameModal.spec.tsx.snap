// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSampleNameModal component test Should render correctly 1`] = `null`;
